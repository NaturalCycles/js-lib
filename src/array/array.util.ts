import { _isNotNullish } from '../is.util'
import type { FalsyValue, Mapper, Predicate, StringMap } from '../types'

/**
 * Creates an array of elements split into groups the length of size. If collection can’t be split evenly, the
 * final chunk will be the remaining elements.
 *
 * @param array The array to process.
 * @param size The length of each chunk.
 * @return Returns the new array containing chunks.
 *
 * https://lodash.com/docs#chunk
 *
 * Based on: https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_chunk
 */
export function _chunk<T>(array: readonly T[], size = 1): T[][] {
  return array.reduce((arr, item, idx) => {
    return idx % size === 0 ? [...arr, [item]] : [...arr.slice(0, -1), [...arr.slice(-1)[0]!, item]]
  }, [] as T[][])
}

/**
 * Removes duplicates from given array.
 */
export function _uniq<T>(a: readonly T[]): T[] {
  return [...new Set(a)]
}

/**
 * This method is like `_.uniq` except that it accepts `iteratee` which is
 * invoked for each element in `array` to generate the criterion by which
 * uniqueness is computed. The iteratee is invoked with one argument: (value).
 *
 * @returns Returns the new duplicate free array.
 * @example
 *
 * _.uniqBy([2.1, 1.2, 2.3], Math.floor);
 * // => [2.1, 1.2]
 *
 * // using the `_.property` iteratee shorthand
 * _.uniqBy([{ 'x': 1 }, { 'x': 2 }, { 'x': 1 }], 'x');
 * // => [{ 'x': 1 }, { 'x': 2 }]
 *
 * Based on: https://stackoverflow.com/a/40808569/4919972
 */
export function _uniqBy<T>(arr: readonly T[], mapper: Mapper<T, any>): T[] {
  return [
    ...arr
      .reduce((map, item, index) => {
        const key = item === null || item === undefined ? item : mapper(item, index)

        if (!map.has(key)) map.set(key, item)

        return map
      }, new Map())
      .values(),
  ]
}

/**
 * const a = [
 *  {id: 'id1', a: 'a1'},
 *  {id: 'id2', b: 'b1'},
 * ]
 *
 * _by(a, r => r.id)
 * // => {
 *   id1: {id: 'id1', a: 'a1'},
 *   id2: {id: 'id2', b: 'b1'},
 * }
 *
 * _by(a, r => r.id.toUpperCase())
 * // => {
 *   ID1: {id: 'id1', a: 'a1'},
 *   ID2: {id: 'id2', b: 'b1'},
 * }
 *
 * Returning `undefined` from the Mapper will EXCLUDE the item.
 */
export function _by<T>(items: readonly T[], mapper: Mapper<T, any>): StringMap<T> {
  return items.reduce((map, item, index) => {
    const res = mapper(item, index)
    if (res !== undefined) map[res] = item
    return map
  }, {} as StringMap<T>)
}

/**
 * const a = [1, 2, 3, 4, 5]
 *
 * _groupBy(a, r => r % 2 ? 'even' : 'odd')
 * // => {
 *   odd: [1, 3, 5],
 *   even: [2, 4],
 * }
 *
 * Returning `undefined` from the Mapper will EXCLUDE the item.
 */
export function _groupBy<T>(items: readonly T[], mapper: Mapper<T, any>): StringMap<T[]> {
  return items.reduce((map, item, index) => {
    const res = mapper(item, index)
    if (res !== undefined) {
      map[res] = [...(map[res] || []), item]
    }
    return map
  }, {} as StringMap<T[]>)
}

/**
 * _sortBy([{age: 20}, {age: 10}], 'age')
 * // => [{age: 10}, {age: 20}]
 *
 * Same:
 * _sortBy([{age: 20}, {age: 10}], o => o.age)
 */
export function _sortBy<T>(
  items: T[],
  mapper: Mapper<T, any>,
  mutate = false,
  descending = false,
): T[] {
  const mod = descending ? -1 : 1
  return (mutate ? items : [...items]).sort((_a, _b) => {
    const [a, b] = [_a, _b].map(mapper)
    if (typeof a === 'number' && typeof b === 'number') return (a - b) * mod
    return String(a).localeCompare(String(b)) * mod
  })
}

/**
 * Like items.find(), but it tries to find from the END of the array.
 */
export function _findLast<T>(items: T[], predicate: Predicate<T>): T | undefined {
  return [...items].reverse().find(predicate)
}

export function _takeWhile<T>(items: T[], predicate: Predicate<T>): T[] {
  let proceed = true
  return items.filter((v, index) => (proceed &&= predicate(v, index)))
}

export function _takeRightWhile<T>(items: T[], predicate: Predicate<T>): T[] {
  let proceed = true
  return [...items].reverse().filter((v, index) => (proceed &&= predicate(v, index)))
}

export function _dropWhile<T>(items: T[], predicate: Predicate<T>): T[] {
  let proceed = false
  return items.filter((v, index) => (proceed ||= !predicate(v, index)))
}

export function _dropRightWhile<T>(items: T[], predicate: Predicate<T>): T[] {
  let proceed = false
  return [...items]
    .reverse()
    .filter((v, index) => (proceed ||= !predicate(v, index)))
    .reverse()
}

export function _countBy<T>(items: T[], mapper: Mapper<T, any>): StringMap<number> {
  const map: StringMap<number> = {}

  items.forEach((item, index) => {
    const key = mapper(item, index)
    map[key] = (map[key] || 0) + 1
  })

  return map
}

// investigate: _groupBy

/**
 * @example
 * _intersection([2, 1], [2, 3])
 * // [2]
 */
export function _intersection<T>(...arrays: T[][]): T[] {
  if (arrays.length === 0) return [] // edge case
  return arrays.reduce((a, b) => a.filter(v => b.includes(v)))
}

/**
 * @example
 * _difference([2, 1], [2, 3])
 * // [1]
 */
export function _difference<T>(source: T[], ...diffs: T[][]): T[] {
  return diffs.reduce((a, b) => a.filter(c => !b.includes(c)), source)
}

/**
 * Returns the sum of items, or 0 for empty array.
 */
export function _sum(items: number[]): number {
  return items.reduce((sum, n) => sum + n, 0)
}

export function _sumBy<T>(items: T[], mapper: Mapper<T, number | undefined>): number {
  return items
    .map((v, i) => mapper(v, i)!)
    .filter(i => typeof i === 'number') // count only numbers, nothing else
    .reduce((sum, n) => sum + n, 0)
}

/**
 * Map an array of T to a StringMap<V>,
 * by returning a tuple of [key, value] from a mapper function.
 * Return undefined/null/false/0/void to filter out (not include) a value.
 *
 * @example
 *
 * _mapToObject([1, 2, 3], n => [n, n * 2])
 * // { '1': 2, '2': 4, '3': 6 }
 *
 * _mapToObject([1, 2, 3], n => [n, `id${n}`])
 * // { '1': 'id1, '2': 'id2', '3': 'id3' }
 */
export function _mapToObject<T, V>(
  array: T[],
  mapper: (item: T) => [key: any, value: V] | FalsyValue,
): StringMap<V> {
  const m: StringMap<V> = {}

  array.forEach(item => {
    const r = mapper(item)
    if (!r) return // filtering

    m[r[0]] = r[1]
  })

  return m
}

/**
 * Randomly shuffle an array values.
 * Fisher–Yates algorithm.
 * Based on: https://stackoverflow.com/a/12646864/4919972
 */
export function _shuffle<T>(array: T[], mutate = false): T[] {
  const a = mutate ? array : [...array]

  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[a[i], a[j]] = [a[j]!, a[i]!]
  }

  return a
}

/**
 * Returns last item of non-empty array.
 * Throws if array is empty.
 */
export function _last<T>(array: T[]): T {
  if (!array.length) throw new Error('_last called on empty array')
  return array[array.length - 1]!
}

/**
 * Returns last item of the array (or undefined if array is empty).
 */
export function _lastOrUndefined<T>(array: T[]): T | undefined {
  return array[array.length - 1]
}

export function _minOrUndefined<T>(array: T[]): NonNullable<T> | undefined {
  const a = array.filter(_isNotNullish)
  if (!a.length) return
  return _min(a)
}

/**
 * Filters out nullish values (undefined and null).
 */
export function _min<T>(array: T[]): NonNullable<T> {
  const a = array.filter(_isNotNullish)
  if (!a.length) throw new Error('_min called on empty array')
  return a.reduce((min, item) => (min <= item ? min : item))
}

export function _maxOrUndefined<T>(array: T[]): NonNullable<T> | undefined {
  const a = array.filter(_isNotNullish)
  if (!a.length) return
  return _max(a)
}

/**
 * Filters out nullish values (undefined and null).
 */
export function _max<T>(array: T[]): NonNullable<T> {
  const a = array.filter(_isNotNullish)
  if (!a.length) throw new Error('_max called on empty array')
  return a.reduce((max, item) => (max >= item ? max : item))
}

export function _maxBy<T>(array: T[], mapper: Mapper<T, number | undefined>): T {
  const max = _maxByOrUndefined(array, mapper)
  if (max === undefined) throw new Error(`_maxBy returned undefined`)
  return max
}

export function _minBy<T>(array: T[], mapper: Mapper<T, number | undefined>): T {
  const min = _minByOrUndefined(array, mapper)
  if (min === undefined) throw new Error(`_minBy returned undefined`)
  return min
}

// todo: looks like it _maxByOrUndefined/_minByOrUndefined can be DRYer

export function _maxByOrUndefined<T>(
  array: T[],
  mapper: Mapper<T, number | undefined>,
): T | undefined {
  if (!array.length) return
  let maxItem: T | undefined
  let max: number | undefined
  array.forEach((item, i) => {
    const v = mapper(item, i)
    if (v !== undefined && (max === undefined || v > max)) {
      maxItem = item
      max = v
    }
  })

  return maxItem
}

export function _minByOrUndefined<T>(
  array: T[],
  mapper: Mapper<T, number | undefined>,
): T | undefined {
  if (!array.length) return
  let minItem: T | undefined
  let min: number | undefined
  array.forEach((item, i) => {
    const v = mapper(item, i)
    if (v !== undefined && (min === undefined || v < min)) {
      minItem = item
      min = v
    }
  })

  return minItem
}
