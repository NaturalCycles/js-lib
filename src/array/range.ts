import { Iterable2 } from '../iter/iterable2'

/* eslint-disable no-redeclare, unicorn/no-new-array */

/**
 * Returns an array with ranges from `from` up to (but not including) `to`.
 *
 * Right bound is Exclusive (not Inclusive), to comply with lodash _.range
 *
 * @example
 * range(3) // [0, 1, 2]
 * range(3, 6) // [ 3, 4, 5 ]
 * range(1, 10, 2) // [ 1, 3, 5, 7, 9 ]
 */
export function _range(toExcl: number): number[]
export function _range(fromIncl: number, toExcl: number, step?: number): number[]
export function _range(fromIncl: number, toExcl?: number, step = 1): number[] {
  if (toExcl === undefined) {
    return Array.from(new Array(fromIncl), (_, i) => i)
  }

  return Array.from(
    { length: Math.ceil((toExcl - fromIncl) / step) },
    (_, i) => i * step + fromIncl,
  )
}

/**
 * Like _range, but returns an Iterable2.
 */
export function _rangeIt(toExcl: number): Iterable2<number>
export function _rangeIt(fromIncl: number, toExcl: number, step?: number): Iterable2<number>
export function _rangeIt(fromIncl: number, toExcl?: number, step = 1): Iterable2<number> {
  if (toExcl === undefined) {
    toExcl = fromIncl
    fromIncl = 0
  }

  return Iterable2.of({
    *[Symbol.iterator]() {
      for (let i = fromIncl; i < toExcl!; i += step) {
        yield i
      }
    },
  })
}
