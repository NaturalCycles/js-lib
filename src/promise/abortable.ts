import { AnyFunction } from '../types'

/**
 * Similar to AbortController and AbortSignal.
 * Similar to pDefer and Promise.
 * Similar to Subject and Observable.
 *
 * Minimal interface for something that can be aborted in the future,
 * but not necessary.
 * Allows to listen to `onAbort` event.
 *
 * @experimental
 */
export class Abortable {
  constructor(public onAbort?: AnyFunction) {}

  aborted = false

  abort(): void {
    if (this.aborted) return
    this.aborted = true
    this.onAbort?.()
    this.onAbort = undefined // cleanup listener
  }

  clear(): void {
    this.onAbort = undefined
  }
}

// convenience function
export function abortable(onAbort?: AnyFunction): Abortable {
  return new Abortable(onAbort)
}
