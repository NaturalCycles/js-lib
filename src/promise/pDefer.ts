/**
 * Similar to Deferred object, which is also a promise itself (instead of deferred.promise).
 */
export interface DeferredPromise<T = void> extends Promise<T> {
  resolve(a?: T): void
  reject(e?: Error): void
}

/* eslint-disable @typescript-eslint/promise-function-async */

/**
 * Returns DeferredPromise - a Promise that has .resolve() and .reject() methods.
 */
export function pDefer<T = void>(): DeferredPromise<T> {
  let resolve: any
  let reject: any

  const promise = new Promise<T>((_resolve, _reject) => {
    resolve = _resolve
    reject = _reject
  }) as DeferredPromise<T>

  promise.resolve = resolve
  promise.reject = reject

  return promise
}
