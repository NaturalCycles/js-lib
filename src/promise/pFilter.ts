import type { AsyncPredicate } from '../types.js'

export async function pFilter<T>(iterable: Iterable<T>, filterFn: AsyncPredicate<T>): Promise<T[]> {
  const items = [...iterable]
  const predicates = await Promise.all(items.map((item, i) => filterFn(item, i)))
  return items.filter((_item, i) => predicates[i])
}
