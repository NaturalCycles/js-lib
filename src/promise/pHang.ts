/**
 * Returns Promise that never resolves ("hanging").
 */
export async function pHang(): Promise<never> {
  return await new Promise<never>(() => void 0)
}
