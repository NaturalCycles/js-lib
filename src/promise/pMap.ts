import type { AbortableAsyncMapper, CommonLogger } from '..'
import { END, ErrorMode, SKIP } from '..'

export interface PMapOptions {
  /**
   * Number of concurrently pending promises returned by `mapper`.
   *
   * Defaults to Infitity.
   */
  concurrency?: number

  /**
   * @default ErrorMode.THROW_IMMEDIATELY
   *
   * When set to THROW_AGGREGATED, instead of stopping when a promise rejects, it will wait for all the promises to settle and then reject with an Aggregated error.
   *
   * When set to SUPPRESS - will ignore errors and return results from successful operations.
   * When in SUPPRESS - errors are still logged via the `logger` (which defaults to console).
   */
  errorMode?: ErrorMode

  /**
   * Default to `console`.
   * Pass `null` to skip logging completely.
   */
  logger?: CommonLogger | null
}

/**
 * Forked from https://github.com/sindresorhus/p-map
 *
 * Improvements:
 * - Exported as { pMap }, so IDE auto-completion works
 * - Included Typescript typings (no need for @types/p-map)
 * - Compatible with pProps (that had typings issues)
 * - Preserves async stack traces (in selected cases)
 *
 * Returns a `Promise` that is fulfilled when all promises in `input` and ones returned from `mapper` are fulfilled,
 * or rejects if any of the promises reject. The fulfilled value is an `Array` of the fulfilled values returned
 * from `mapper` in `input` order.
 *
 * @param iterable - Iterated over concurrently in the `mapper` function.
 * @param mapper - Function which is called for every item in `input`. Expected to return a `Promise` or value.
 * @param opt - Options-object.
 *
 * @example
 *
 * const sites = [
 * 	'ava.li',
 * 	'todomvc.com
 * ];
 *
 * (async () => {
 * 	const mapper = async site => {
 * 		const {requestUrl} = await got.head(site);
 * 		return requestUrl;
 * 	};
 *
 * 	const result = await pMap(sites, mapper, {concurrency: 2});
 * 	//=> ['http://ava.li/', 'http://todomvc.com/']
 * })();
 */
export async function pMap<IN, OUT>(
  iterable: Iterable<IN>,
  mapper: AbortableAsyncMapper<IN, OUT>,
  opt: PMapOptions = {},
): Promise<OUT[]> {
  const items = [...iterable]
  const itemsLength = items.length
  if (itemsLength === 0) return [] // short circuit

  const { concurrency = Infinity, errorMode = ErrorMode.THROW_IMMEDIATELY, logger = console } = opt

  // Special cases that are able to preserve async stack traces
  // Special case: serial execution
  if (concurrency === 1) {
    return await pMap1(items, mapper, errorMode, logger)
  }

  // Special case: items.length <= concurrency (including when concurrency === Infinity)
  if (items.length <= concurrency) {
    return await pMapAll(items, mapper, errorMode, logger)
  }

  // General case: execution with throttled concurrency
  const ret: (OUT | typeof SKIP)[] = []
  const errors: Error[] = []
  let isSettled = false
  let resolvingCount = 0
  let currentIndex = 0

  return await new Promise<OUT[]>((resolve, reject) => {
    const next = (): void => {
      if (isSettled) {
        return
      }

      const nextItem = items[currentIndex]!
      const i = currentIndex++

      if (currentIndex > itemsLength) {
        if (resolvingCount === 0) {
          isSettled = true
          const r = ret.filter(r => r !== SKIP) as OUT[]
          if (errors.length) {
            reject(new AggregateError(errors, `pMap resulted in ${errors.length} error(s)`))
          } else {
            resolve(r)
          }
        }

        return
      }

      resolvingCount++

      Promise.resolve(nextItem)
        .then(async element => await mapper(element, i))
        .then(
          value => {
            if (value === END) {
              isSettled = true
              return resolve(ret.filter(r => r !== SKIP) as OUT[])
            }

            ret[i] = value
            resolvingCount--
            next()
          },
          (err: Error) => {
            if (errorMode === ErrorMode.THROW_IMMEDIATELY) {
              isSettled = true
              reject(err)
            } else {
              if (errorMode === ErrorMode.THROW_AGGREGATED) {
                errors.push(err)
              } else {
                // otherwise, suppress (but still log via logger)
                logger?.error(err)
              }
              resolvingCount--
              next()
            }
          },
        )
    }

    for (let i = 0; i < concurrency; i++) {
      next()

      if (isSettled) {
        break
      }
    }
  })
}

/**
 pMap with serial (non-concurrent) execution.
 */
async function pMap1<IN, OUT>(
  items: IN[],
  mapper: AbortableAsyncMapper<IN, OUT>,
  errorMode: ErrorMode,
  logger: CommonLogger | null,
): Promise<OUT[]> {
  let i = 0
  const ret: OUT[] = []
  const errors: Error[] = []

  for (const item of items) {
    try {
      const r = await mapper(item, i++)
      if (r === END) break
      if (r !== SKIP) ret.push(r)
    } catch (err) {
      if (errorMode === ErrorMode.THROW_IMMEDIATELY) throw err
      if (errorMode === ErrorMode.THROW_AGGREGATED) {
        errors.push(err as Error)
      } else {
        // otherwise, suppress (but still log via logger)
        logger?.error(err)
      }
    }
  }

  if (errors.length) {
    throw new AggregateError(errors, `pMap resulted in ${errors.length} error(s)`)
  }

  return ret
}

/**
 pMap with fully concurrent execution, like Promise.all
 */
async function pMapAll<IN, OUT>(
  items: IN[],
  mapper: AbortableAsyncMapper<IN, OUT>,
  errorMode: ErrorMode,
  logger: CommonLogger | null,
): Promise<OUT[]> {
  if (errorMode === ErrorMode.THROW_IMMEDIATELY) {
    return (await Promise.all(items.map((item, i) => mapper(item, i)))).filter(
      r => r !== SKIP && r !== END,
    ) as OUT[]
  }

  const ret: OUT[] = []
  const errors: Error[] = []

  for (const r of await Promise.allSettled(items.map((item, i) => mapper(item, i)))) {
    if (r.status === 'fulfilled') {
      if (r.value === END) break
      if (r.value !== SKIP) ret.push(r.value)
    } else if (errorMode === ErrorMode.THROW_AGGREGATED) {
      errors.push(r.reason)
    } else {
      // otherwise, suppress (but still log via logger)
      logger?.error(r.reason)
    }
  }

  if (errors.length) {
    throw new AggregateError(errors, `pMap resulted in ${errors.length} error(s)`)
  }

  return ret
}
