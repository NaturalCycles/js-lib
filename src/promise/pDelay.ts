import type { PromisableFunction } from '../types'
import { DeferredPromise, pDefer } from './pDefer'

/**
 * Promisified version of setTimeout.
 *
 * Can return a value.
 * If value is instanceof Error - rejects the Promise instead of resolving.
 */
export async function pDelay<T>(ms = 0, value?: T): Promise<T> {
  return await new Promise<T>((resolve, reject) =>
    setTimeout(value instanceof Error ? reject : resolve, ms, value),
  )
}

/* eslint-disable @typescript-eslint/promise-function-async */

/**
 * Promisified version of setTimeout.
 *
 * Wraps the passed function with try/catch,
 * catch will propagate to pDelayFn rejection,
 * otherwise pDelayFn will resolve with returned value.
 *
 * On abort() - clears the Timeout and immediately resolves the Promise with void.
 */
export function pDelayFn<T>(ms = 0, fn: PromisableFunction<T>): DeferredPromise<T> {
  const p = pDefer<T>()

  const timer = setTimeout(async () => {
    try {
      p.resolve(await fn())
    } catch (err) {
      p.reject(err as Error)
    }
  }, ms)

  p.abort = () => {
    clearTimeout(timer)
    // p.rejectAborted(reason) // nope
    p.resolve()
  }

  return p
}
