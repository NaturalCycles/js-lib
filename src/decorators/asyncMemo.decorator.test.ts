import { _range } from '../array/range'
import { pDelay } from '../promise/pDelay'
import { _AsyncMemo, _getAsyncMemo } from './asyncMemo.decorator'
import { MapAsyncMemoCache } from './memo.util'

class A {
  func(n: number): void {
    console.log(`func ${n}`)
  }

  @_AsyncMemo({ cacheFactory: () => new MapAsyncMemoCache() })
  async a(a1: number, a2: number): Promise<number> {
    const n = a1 * a2
    this.func(n)
    return n
  }
}

test('memo a', async () => {
  const a = new A()
  jest.spyOn(a, 'func').mockImplementation()

  // first call
  let r = await a.a(2, 3)
  expect(r).toBe(6)
  r = await a.a(3, 4)
  expect(r).toBe(12)

  // second call
  r = await a.a(2, 3)
  expect(r).toBe(6)
  r = await a.a(3, 4)
  expect(r).toBe(12)

  // to be called once per set of arguments (2)
  expect(a.func).toMatchSnapshot()

  // cleanup for the next tests
  await _getAsyncMemo(a.a).clear()
})

test('MEMO_DROP_CACHE', async () => {
  const a = new A()
  jest.spyOn(a, 'func').mockImplementation()

  // first call
  await a.a(2, 3)

  // drop cache
  await _getAsyncMemo(a.a).clear()
  expect(_getAsyncMemo(a.a).getInstanceCache().size).toBe(0)

  // second call
  await a.a(2, 3)

  expect(a.func).toHaveBeenCalledTimes(2)
})

class B {
  cacheMisses = 0

  @_AsyncMemo({ cacheFactory: () => new MapAsyncMemoCache() }) // testing 2 layers of AsyncMemo!
  @_AsyncMemo({
    // testing to provide specific cacheFactory, should be no difference in this test
    cacheFactory: () => new MapAsyncMemoCache(),
  })
  async a(a1 = 'def'): Promise<number> {
    console.log(`a called with a1=${a1}`)
    this.cacheMisses++
    return 1 // it cannot return undefined, otherwise it'll always MISS
  }
}

test('should work with default arg values', async () => {
  const b = new B()

  // Call 1 with default (cache miss). arg1=undefined
  await b.a()

  // Call 2 with default (cache hit). arg1=undefined
  await b.a()

  // Call 3 with same as default (cache miss). arg1=def
  await b.a('def')

  // Call 4 with non-default (cache miss). arg1=nondef
  await b.a('nondef')

  expect(b.cacheMisses).toBe(3)
})

class C {
  calls = 0

  @_AsyncMemo({
    cacheFactory: () => new MapAsyncMemoCache(10),
  })
  async fn(): Promise<void> {
    this.calls++
    await pDelay(100)
  }
}

test('swarm test of async function', async () => {
  // Swarm is when one "slow" async function is called multiple times in parallel
  // Expectation is that it should return the same Promise and `calls` should equal to 1
  // Because it **synchronously** returns a Promise (despite the fact that it's async)
  const c = new C()

  await Promise.all(_range(10).map(() => c.fn()))

  expect(c.calls).toBe(1)

  await pDelay(150)
  await Promise.all(_range(3).map(() => c.fn()))
  expect(c.calls).toBe(1)
})
