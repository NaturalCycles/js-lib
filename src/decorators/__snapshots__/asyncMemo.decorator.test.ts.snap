// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`memo a 1`] = `
[MockFunction func] {
  "calls": [
    [
      6,
    ],
    [
      12,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
